---
title: "Inversion VCF"
author: "KE Lotterhos"
date: "1/4/2021"
output: html_document
---

IF MAF < 0.05 then FST has 0 value


```{r setup, include=FALSE}

packages_needed <- c("IntegratedMRF", "vcfR", "distances","ggplot2", "metR", "fields", "MultivariateRandomForest", "akima", "MLmetrics", "ash", "plotly", "stringr")

for (i in 1:length(packages_needed)){
  if(!(packages_needed[i] %in% installed.packages())){install.packages(packages_needed[i])}
}

#  require(ggplot2)}

for (i in 1:length(packages_needed)){
  library( packages_needed[i], character.only = TRUE)
}
```

```{r}
seed <- 3384725
path <- paste0("../",seed,"_Movie/")

# Individuals and phenotypes
indPhen_df<- read.table(paste0(path,seed,"_outputIndPheno.txt"), header=TRUE)
dim(indPhen_df)
head(indPhen_df)

# Mutation stats at end of sim
# for MAF > 0.01?
muts_df <- read.table(paste0(path,seed,"_outputMutations.txt"), header=TRUE)
dim(muts_df)
head(muts_df)
table(muts_df$type) #just m2 mutations
hist(muts_df$FST)
hist(muts_df$freq)

# inversion summary
inv_df <- read.table(paste0(path,seed,"_outputInvSumInfo.txt"), header=TRUE)
head(inv_df)

# VCF file
vcffile <- list.files(path=path, pattern=paste0(".vcf"))
vcf <- read.vcfR(paste0(path,vcffile))
head(vcf)
head(vcf@fix, 50)
dim(vcf@fix)
# example of how to find a specific mutation in the vcf file
  muts_df[2,]
  vcf@fix[grep(muts_df$mutID[1], vcf@fix[,"INFO"]),]
```

```{r}
 boxplot(indPhen_df$fitness~indPhen_df$subpop, ylab = "fitness", xlab="pop", main="all", ylim=c(0,1))

  boxplot(indPhen_df$phenotype~indPhen_df$subpop, ylim=c(-2,2))
  abline(h=0, col="blue")
```

```{r vcf}
head(muts_df)
dim(vcf@gt)
  # rows are mutations
  # columns are individuals

head(vcf@gt[,1:5])
head(vcf@gt[,190:195])

head(vcf@fix)

geno <- vcf@gt[,-1] 
position <- getPOS(vcf)
if (sum(duplicated(position)) != 0){
  print("This simulation needs to be checked for duplicated locus positions")
}

G <- matrix(NA, nrow = nrow(geno), ncol = ncol(geno))
G[geno %in% c("0/0", "0|0")] <- 0
G[geno  %in% c("0/1", "1/0", "1|0", "0|1")] <- 1
G[geno %in% c("1/1", "1|1")] <- 2

a_freq <- rowSums(G)/(2*ncol(G))
hist(a_freq)

# keepmuts <- which(a_freq>0.01) # already filtered VCF file

# Individuals in VCF
vcf_ind <- data.frame(vcf_ind=colnames(vcf@gt)[-1])

colnames(G) <- vcf_ind$vcf_ind
head(G[,1:5])
dim(G)

dim(vcf_ind)
head(vcf_ind)
head(indPhen_df)
indPhen_df$vcf_ind <- paste0("i",0:1999) # hard coding
# The individual IDs in Slim do not match the IDs in the VCF file. 
# I will assume they are in the same order
tail(indPhen_df)

indPhen_df_vcf <- merge(vcf_ind, indPhen_df, by="vcf_ind")
dim(indPhen_df)
dim(indPhen_df_vcf)

indPhen_df_vcf <- indPhen_df_vcf[order(indPhen_df_vcf$subpop, indPhen_df_vcf$id),]
head(indPhen_df_vcf)
tail(indPhen_df_vcf)
```

## individual order
```{r}
## Plot population 1 with individuals sorted
pop1_ids <- which(indPhen_df_vcf$subpop==1)
pop2_ids <- which(indPhen_df_vcf$subpop==2)

G_pop1 <- G[, pop1_ids]
G_pop2 <- G[, pop2_ids]
dim(G_pop2)
head(G_pop1[,1:5])
head(G_pop2[,1:5])

## Population 1
  fordist <- as.data.frame(t(G_pop1))
  dist_mat <- dist(fordist, method="euclidean")
  dim(dist_mat)
  pop1_clust <- hclust(dist_mat, method = "ward.D")
  str(pop1_clust)
  
  pop1_order <- pop1_clust$order
  
  heatmap(t(G_pop1[,pop1_order]), Rowv = NA,  main="Pop1 genotypes",cexCol = 0.3,
             Colv = NA,
             labRow = indPhen_df_vcf$subpop)

## Population 2
  fordist <- as.data.frame(t(G_pop2))
  dist_mat <- dist(fordist, method="euclidean")
  dim(dist_mat)
  pop2_clust <- hclust(dist_mat, method = "ward.D")
  str(pop2_clust)
  
  pop2_order <- pop2_clust$order
  
  heatmap(t(G_pop2[,pop2_order]), Rowv = NA,  main="Pop2 genotypes",cexCol = 0.3,
             Colv = NA,
             labRow = indPhen_df_vcf$subpop)
```

# TO DO: need to change row labels to be chromosome posiiton

This chunk has some other heatmapping. I realized that the heatmapping clustering really sucked, so I used my own order from hclust.

```{r, echo=FALSE, eval=FALSE}

dim(G) # columns are individuals

## Plot raw data
a<-heatmap(t(G), Rowv = NA,  main="All genotypes",cexCol = 0.3,
           Colv = NA,
           labRow = indPhen_df_vcf$subpop)

heatmap(t(G[, pop1_ids]),   main="Pop1 genotypes",cexCol = 0.3,
           Colv = NA,
        # Rowv = NA,
           labRow = 1)

heatmap(t(G[, pop2_ids]),   main="Pop2 genotypes",cexCol = 0.3,
           Colv = NA,
        # Rowv = NA,
           labRow = 2)

```

## Heatmaps based on contribution to phenotype

Each allele has an additive effect on the phenotype, which is stored in the `vcf@fix[,"INFO"]` as S. IF the mutation type is 2 (MT=2), the additive effect of an allele on that individual's phenotype is S*(the number of alleles).

For neutral mutations (MT=1) and inversion mutations (MT=3), S=0, so they won't affect the calculation.
```{r}
head(vcf@fix[,"INFO"])
tail(vcf@fix[,"INFO"])

whichinversionmuts <- grep("MT=3", vcf@fix[,"INFO"]) #inversions

vcf@fix[whichinversionmuts,"INFO"]
info <- str_split(vcf@fix[,"INFO"], pattern =";", simplify=TRUE)
# This took like an hour to figure out
a<- as.numeric(substring(info[,2], first=3))
head(a)
hist(a, breaks=seq(-0.01, 0.01, length.out=101))
summary(a)
length(a)
dim(G)

dim(G_pop1)
head(G_pop1[,1:10])

G1_alpha <- G_pop1*a # make sure G and a line up
head(G1_alpha[,1:50])

#G1_alpha[,1]

hist(G1_alpha, breaks=seq(-0.02, 0.02, length.out=101))

G2_alpha <- G_pop2*a # make sure G and a line up
head(G2_alpha[,1:10])
hist(G2_alpha, breaks=seq(-0.02, 0.02, length.out=101))

hist(colSums(G1_alpha))
hist(colSums(G2_alpha))

# Sanity check - mutations in rows
head(G[1:100,1:10])
t(G1_alpha[1:100,9:10])

hist(G1_alpha)
hist(G2_alpha)
dim(G1_alpha)
heatmap(t(G1_alpha[1:1000, pop1_order]),   
        main="Pop1 G*a",cexCol = 0.3,
           Colv = NA,
        # Rowv = NA, 
       col=two.colors(100, start = "blue", end="red", middle="white") )
      # ADDING BREAKS SCREWS UP EVERYTHING
        #breaks=seq(-0.005, 0.005, length.out = 101))

heatmap(t(G2_alpha[1:1000, pop2_order]),   
        main="Pop2 G*a",cexCol = 0.3,
           Colv = NA,
       #  Rowv = NA, 
       col=two.colors(100, start = "blue", end="red", middle="white") )
```

# line up positions
```{r}
vcf_pos <- as.numeric(vcf@fix[,"POS"])
hist(vcf_pos, breaks=seq(0,2100000, length.out=100))
hist(muts_df$position, breaks=seq(0,2100000, length.out=100))

head(sort(muts_df$position))
head(vcf_pos)
muts_df$position_vcf <- muts_df$position + 1
muts_df$is_vcf <- NA

G_FST <- rep(NA, nrow(G))

# this is slow, but correct
for (i in 1:nrow(muts_df)){
  x<- grep(muts_df$mutID[i], vcf@fix[,"INFO"])
  if (length(x)==1){
    G_FST[x] <- muts_df$FST[i]
  }
}

hist(muts_df$FST)
hist(G_FST)

sum(is.na(G_FST))
# mutations that don't match up

length(a)
length(G_FST)
```

# G x a x FST
```{r}
head(G[,1:5])
# Allele freq is sum of row / 2N

p <- rowSums(G)/400 # hard coding
hist(p)


G1_alpha <- G_pop1*a*G_FST # make sure G and a line up
G2_alpha <- G_pop2*a*G_FST # make sure G and a line up


# Sanity check - mutations in rows

heatmap(t(G1_alpha[ ,pop1_order]),   
        main="Pop1 G*a*FST",cexCol = 0.3,
           Colv = NA,
        # Rowv = NA, 
       col=two.colors(100, start = "blue", end="red", middle="white") )
      # ADDING BREAKS SCREWS UP EVERYTHING
        #breaks=seq(-0.005, 0.005, length.out = 101))

heatmap(t(G2_alpha[, pop2_order]),   
        main="Pop2 G*a*FST",cexCol = 0.3,
           Colv = NA,
       #  Rowv = NA, 
       col=two.colors(100, start = "blue", end="red", middle="white") )
```


# (allele count) x (+ or - effect on phenotype) x FST

```{r}
hist(a)
a2 <- a
a2[a>0.001] <- 1
a2[a<0.001] <- -1


G1_alpha <- G_pop1*a2*G_FST # make sure G and a line up
G2_alpha <- G_pop2*a2*G_FST # make sure G and a line up

hist(G_pop1*a2)

# Sanity check - mutations in rows

heatmap(t(G1_alpha[, pop1_order]),   
        main="Pop1 G*a+-*FST",cexCol = 0.3,
           Colv = NA,
        # Rowv = NA, 
       col=two.colors(100, start = "blue", end="red", middle="white") )
      # ADDING BREAKS SCREWS UP EVERYTHING
        #breaks=seq(-0.005, 0.005, length.out = 101))

heatmap(t(G2_alpha[, pop2_order]),   
        main="Pop2 G*a+-*FST",cexCol = 0.3,
           Colv = NA,
       #  Rowv = NA, 
       col=two.colors(100, start = "blue", end="red", middle="white") )

```


```{r}
plot(a^2*p*(1-p), G_FST)
#inv_df
```



# percent of additive genetic variance in inversions
```{r}
head(info)
dim(info)

# Inversion locations
grep("MT=3", vcf@fix[,"INFO"])
grep("MT=3", info[,6])

inv_id <- data.frame(inv_id=as.numeric(substring(info[ grep("MT=3", info[,6]), 1], first=5)))

inv_df_sm <- merge(inv_id, inv_df)

is.window <- function(pos){
  sum(inv_df_sm$inv_pos < pos & pos < inv_df_sm$inv_end)>0
}
is.window(604531)
is.window(604533)

G_isInvWindow <- rep(NA, nrow(G))
for (i in 1:nrow(G)){
  G_isInvWindow[i] <- is.window(as.numeric(vcf@fix[i,"POS"]))
}

Va <- a^2*p*(1-p)
Va_perc <- Va/sum(Va)

sum(Va_perc[G_isInvWindow])
sum(Va_perc[!G_isInvWindow])

boxplot(G_FST~G_isInvWindow)

# Percent of genome covered by inversions
sum(inv_df_sm$inv_end - inv_df_sm$inv_pos)/max(muts_df$position)

```

# TO DO:
Add color of individuals fitness on right side

Add blocks on top showing where inversions are, use alpha so we can look for overlaps

Check Sara's FST output matches VCF FST output (?)