---
title: "Inversion VCF"
author: "KE Lotterhos"
date: "1/4/2021"
output: html_document
---

```{r setup, include=FALSE}

packages_needed <- c("IntegratedMRF", "vcfR", "distances","ggplot2", "metR", "fields", "MultivariateRandomForest", "akima", "MLmetrics", "ash", "plotly", "stringr")

for (i in 1:length(packages_needed)){
  if(!(packages_needed[i] %in% installed.packages())){install.packages(packages_needed[i])}
}

#  require(ggplot2)}

for (i in 1:length(packages_needed)){
  library( packages_needed[i], character.only = TRUE)
}
```

```{r}
seed <- 3384725
path <- paste0("../",seed,"_Movie/")

# Individuals and phenotypes
indPhen_df<- read.table(paste0(path,seed,"_outputIndPheno.txt"), header=TRUE)
dim(indPhen_df)
head(indPhen_df)

# Mutation stats at end of sim
# for MAF > 0.01?
muts_df <- read.table(paste0(path,seed,"_outputMutations.txt"), header=TRUE)
dim(muts_df)
head(muts_df)
table(muts_df$type) #just m2 mutations
hist(muts_df$FST)
hist(muts_df$freq)

# inversion summary
inv_df <- read.table(paste0(path,seed,"_outputInvSumInfo.txt"), header=TRUE)
head(inv_df)

# VCF file
vcffile <- list.files(path=path, pattern=paste0(".vcf"))
vcf <- read.vcfR(paste0(path,vcffile))
head(vcf)
head(vcf@fix, 50)
dim(vcf@fix)
# example of how to find a specific mutation in the vcf file
  muts_df[2,]
  vcf@fix[grep(muts_df$mutID[1], vcf@fix[,"INFO"]),]
```

```{r}
 boxplot(indPhen_df$fitness~indPhen_df$subpop, ylab = "fitness", xlab="pop", main="all", ylim=c(0,1))

  boxplot(indPhen_df$phenotype~indPhen_df$subpop, ylim=c(-2,2))
  abline(h=0, col="blue")
```

```{r vcf}
head(muts_df)
dim(vcf@gt)
  # rows are mutations
  # columns are individuals

head(vcf@gt[,1:5])
head(vcf@gt[,190:195])

head(vcf@fix)

geno <- vcf@gt[,-1] 
position <- getPOS(vcf)
if (sum(duplicated(position)) != 0){
  print("This simulation needs to be checked for duplicated locus positions")
}

G <- matrix(NA, nrow = nrow(geno), ncol = ncol(geno))
G[geno %in% c("0/0", "0|0")] <- 0
G[geno  %in% c("0/1", "1/0", "1|0", "0|1")] <- 1
G[geno %in% c("1/1", "1|1")] <- 2

a_freq <- rowSums(G)/(2*ncol(G))
hist(a_freq)

# keepmuts <- which(a_freq>0.01) # already filtered VCF file

# Individuals in VCF
vcf_ind <- data.frame(vcf_ind=colnames(vcf@gt)[-1])

colnames(G) <- vcf_ind$vcf_ind
head(G[,1:5])
dim(G)

dim(vcf_ind)
head(vcf_ind)
head(indPhen_df)
indPhen_df$vcf_ind <- paste0("i",0:1999) # hard coding
# The individual IDs in Slim do not match the IDs in the VCF file. 
# I will assume they are in the same order
tail(indPhen_df)

indPhen_df_vcf <- merge(vcf_ind, indPhen_df, by="vcf_ind")
dim(indPhen_df)
dim(indPhen_df_vcf)

indPhen_df_vcf <- indPhen_df_vcf[order(indPhen_df_vcf$subpop, indPhen_df_vcf$id),]
head(indPhen_df_vcf)
tail(indPhen_df_vcf)
```


## Heatmaps based on allele frequencies
```{r}

dim(G) # columns are individuals

## Plot raw data
a<-heatmap(t(G), Rowv = NA,  main="All genotypes",cexCol = 0.3,
           Colv = NA,
           labRow = indPhen_df_vcf$subpop)


## Plot population 1 with individuals sorted
pop1_ids <- which(indPhen_df_vcf$subpop==1)
pop2_ids <- which(indPhen_df_vcf$subpop==2)

# TO DO: need to change row labels to be chromosome posiiton
heatmap(t(G[, pop1_ids]),   main="Pop1 genotypes",cexCol = 0.3,
           Colv = NA,
        # Rowv = NA,
           labRow = 1)

heatmap(t(G[, pop2_ids]),   main="Pop2 genotypes",cexCol = 0.3,
           Colv = NA,
        # Rowv = NA,
           labRow = 2)

```

## Heatmaps based on contribution to phenotype

Each allele has an additive effect on the phenotype, which is stored in the `vcf@fix[,"INFO"]` as S. IF the mutation type is 2 (MT=2), the additive effect of an allele on that individual's phenotype is S*(the number of alleles).

For neutral mutations (MT=1) and inversion mutations (MT=3), S=0, so they won't affect the calculation.
```{r}
head(vcf@fix[,"INFO"])
tail(vcf@fix[,"INFO"])

whichinversionmuts <- grep("MT=3", vcf@fix[,"INFO"]) #inversions

vcf@fix[whichinversionmuts,"INFO"]

# This took like an hour to figure out
a<- as.numeric(substring(str_split(vcf@fix[,"INFO"], pattern =";", simplify=TRUE)[,2], first=3))
head(a)
hist(a)
summary(a)
length(a)
dim(G)

G_alpha <- G*a # make sure G and a line up
max(abs(as.numeric(G_alpha)))
hist(G_alpha)
hist(G)

# Sanity check - mutations in rows
head(G[,1:10])


heatmap(t(G_alpha[, pop1_ids]),   main="Pop1 genotypes",cexCol = 0.3,
           Colv = NA,
        # Rowv = NA,
           labRow = 1, 
        col=two.colors(10, start = "blue", end="red", middle="grey"), 
        breaks=seq(-0.005, 0.005, length.out = 11))

heatmap(t(G_alpha[, pop2_ids]),   main="Pop2 genotypes",cexCol = 0.3,
           Colv = NA,
        # Rowv = NA,
           labRow = 2, 
        col=two.colors(100, start = "blue", end="red", middle="grey"), 
        breaks=seq(-0.005, 0.005, length.out = 101))

```

## Same heatmaps but with different breaks

I don't understand why this works
```{r}

heatmap(t(G_alpha[, pop1_ids]),   main="Pop1 genotypes",cexCol = 0.3,
           Colv = NA,
        # Rowv = NA,
           labRow = 1, 
        col=two.colors(100, start = "blue", end="red", middle="grey"), 
        breaks=seq(-0.55, 0.55, length.out = 101))

heatmap(t(G_alpha[, pop2_ids]),   main="Pop2 genotypes",cexCol = 0.3,
           Colv = NA,
        # Rowv = NA,
           labRow = 2, 
        col=two.colors(100, start = "blue", end="red", middle="grey"), 
        breaks=seq(-0.55, 0.55, length.out = 101))
```


# Just  visualize + or - effect on phenotype
```{a}
hist(a)
a2 <- a
a2[a>0.001] <- 1
a2[a<0.001] <- -1

hist(a2)
G3 <- a2*G
head(G3)
hist(G3)

heatmap(t(G3[, pop1_ids]),   main="Pop1 genotypes",cexCol = 0.3,
           Colv = NA,
        # Rowv = NA,
           labRow = 1, 
        col=two.colors(5, start = "blue", end="red", middle="white"), 
        breaks=-2.5:2.5)

heatmap(t(G3[, pop2_ids]),   main="Pop2 genotypes",cexCol = 0.3,
           Colv = NA,
        # Rowv = NA,
           labRow = 2, 
        col=two.colors(5, start = "blue", end="red", middle="white"), 
        breaks=-2.5:2.5)

```

#  order individuals in the heatmap by their fitness

```{r, eval=FALSE}
plot(indPhen_df_vcf$fitness[pop1_ids], rank(indPhen_df_vcf$fitness[pop1_ids]))

pop1_order <- order(indPhen_df_vcf$fitness[pop1_ids], decreasing=TRUE)

head(indPhen_df_vcf)
head(G3[, 1:5])

# SANITY CHECK
head(indPhen_df_vcf[pop1_order,])
head(G3[, pop1_order[1:5]])

heatmap(t(G3[, rev(pop1_order)]),   main="Pop1 genotypes",cexCol = 0.3,
           Colv = NA,
         Rowv = NA,
           labRow = 1, 
        col=two.colors(5, start = "blue", end="red", middle="white"), 
        breaks=-2.5:2.5)
```


# multiply allele effect by p*(1-p) (gets at VA)
```{r}
head(G[,1:5])
# Allele freq is sum of row / 2N

p <- rowSums(G)/400 # hard coding
hist(p)
G4 <- a*G*(p)*(1-p)
hist(G4)

heatmap(t(G4[, pop1_ids]),   main="Pop1 genotypes",cexCol = 0.3,
           Colv = NA,
        # Rowv = NA,
           labRow = 1, 
        col=two.colors(100, start = "blue", end="red", middle="grey"), 
        breaks=seq(-0.55, 0.55, length.out = 101))


heatmap(t(G4[, pop2_ids]),   main="Pop2 genotypes",cexCol = 0.3,
           Colv = NA,
        # Rowv = NA,
           labRow = 1, 
        col=two.colors(100, start = "blue", end="red", middle="grey"), 
        breaks=seq(-0.55, 0.55, length.out = 101))

```

Same heatmaps but with different breaks
```{r}

heatmap(t(G4[, pop1_ids]),   main="Pop1 genotypes",cexCol = 0.3,
           Colv = NA,
        # Rowv = NA,
           labRow = 1, 
        col=two.colors(100, start = "blue", end="red", middle="grey"), 
        breaks=seq(-0.05, 0.05, length.out = 101))


heatmap((G4[, pop2_ids]),   main="Pop2 genotypes",cexCol = 0.3,
        #   Colv = NA,
         Rowv = NA,
           labRow = 1, 
        col=two.colors(100, start = "blue", end="red", middle="grey"), 
        breaks=seq(-0.05, 0.05, length.out = 101))
```

# TO DO:
Add color of individuals fitness on right side
Add blocks on top showing where inversions are, use alpha so we can look for overlaps
```{r}
hist(indPhen_df_vcf$phenotype)
hist(colSums(G))
```