---
title: "Inversion movie"
author: "KE Lotterhos"
date: "1/4/2021"
output: html_document
---

```{r setup, include=FALSE}

packages_needed <- c("IntegratedMRF", "vcfR", "distances","ggplot2", "metR", "fields", "MultivariateRandomForest", "akima", "MLmetrics", "ash", "plotly")

for (i in 1:length(packages_needed)){
  if(!(packages_needed[i] %in% installed.packages())){install.packages(packages_needed[i])}
}

#  require(ggplot2)}

for (i in 1:length(packages_needed)){
  library( packages_needed[i], character.only = TRUE)
}
```

```{r}
seed <- 3384725
path <- paste0("~/Documents/GitHub/CoreSim_QTL/inversion_movie/",seed,"_Movie/")

# Individuals and phenotypes
indPhen_df<- read.table(paste0(path,seed,"_outputIndPheno.txt"), header=TRUE)
dim(indCG_df)
head(indCG_df)

# Mutation stats at end of sim
# for MAF > 0.01?
muts_df <- read.table(paste0(path,seed,"_outputMutations.txt"), header=TRUE)
dim(muts_df)
head(muts_df)
table(muts_df$type) #just m2 mutations
hist(muts_df$FST)
hist(muts_df$freq)

# inversion summary
inv_df <- read.table(paste0(path,seed,"_outputInvSumInfo.txt"), header=TRUE)
head(inv_df)

# VCF file
vcffile <- list.files(path=path, pattern=paste0(".vcf"))
vcf <- read.vcfR(vcffile)
head(vcf)
head(vcf@fix, 50)
dim(vcf@fix)
# example of how to find a specific mutation in the vcf file
  muts_df[2,]
  vcf@fix[grep(muts_df$mutID[1], vcf@fix[,"INFO"]),]
```

```{r}
 boxplot(indPhen_df$fitness~indPhen_df$subpop, ylab = "fitness", xlab="pop", main="all", ylim=c(0,1))

  boxplot(indPhen_df$phenotype~indPhen_df$subpop, ylim=c(-2,2))
  abline(h=0, col="blue")
```

```{r vcf}
head(muts_df)
dim(vcf@gt)
  # rows are mutations
  # columns are individuals

head(vcf@gt[,1:5])
head(vcf@gt[,190:195])

head(vcf@fix)

geno <- vcf@gt[,-1] 
position <- getPOS(vcf)
if (sum(duplicated(position)) != 0){
  print("This simulation needs to be checked for duplicated locus positions")
}

G <- matrix(NA, nrow = nrow(geno), ncol = ncol(geno))
G[geno %in% c("0/0", "0|0")] <- 0
G[geno  %in% c("0/1", "1/0", "1|0", "0|1")] <- 1
G[geno %in% c("1/1", "1|1")] <- 2

a_freq <- rowSums(G)/(2*ncol(G))
hist(a_freq)

# keepmuts <- which(a_freq>0.01) # already filtered VCF file

# Individuals in VCF
vcf_ind <- data.frame(vcf_ind=colnames(vcf@gt)[-1])
dim(vcf_ind)
head(vcf_ind)
head(indPhen_df)
indPhen_df$vcf_ind <- paste0("i",0:1999) # hard coding
# The individual IDs in Slim do not match the IDs in the VCF file. 
# I will assume they are in the same order
tail(indPhen_df)

indPhen_df_vcf <- merge(vcf_ind, indPhen_df, by="vcf_ind")
dim(indPhen_df)
dim(indPhen_df_vcf)

indPhen_df_vcf <- indPhen_df_vcf[order(indPhen_df_vcf$subpop),]
head(indPhen_df_vcf)
tail(indPhen_df_vcf)
```

```{r}

a<-heatmap(t(G), Rowv = NA,  main="All genotypes",cexCol = 0.3,
           #Colv = NA,
           labRow = indPhen_df_vcf$subpop)

```











```{r}

df <- read.table("movieData.txt", header=TRUE)
head(df)
tail(df)
df$chrom = floor(df$inv_pos/100000)
df$chrom_dec = df$inv_pos/100000

hist(df$freq)
df$inv_mid <- (df$inv_pos + df$inv_end)/2
max(df$inv_FST)

min(df$inv_originGen)
max(df$inv_originGen)

all_gen <- unique(df$sim_gen)
max(df_sub$num_qtns)
```

```{r}
k=0
length(all_gen)
for (gen in sort(all_gen)){
  k=k+1
  df_sub <- df %>% filter(sim_gen==gen) %>% arrange(inv_pos)
  df_sub
  
   png(paste0("movie/invmovie_", sprintf("%05.f", gen), ".png"), width=10, height = 8,units = "in", res = 600)
   par(mfrow=c(2,1), mar=c(2,4,0,0), oma=c(3,0,3,0))
   ### Plot 1 ####
  plot(y=c(-0.01,0.25),x=c(0, 2.3e+06) , col=rgb(0,0,0,0),
       ylab="Fst", bty="l", xaxt="n", xlab="",
         )
  abline(h=0)
  if (gen<10000){
    main = paste0("generation ", gen, " (burnin)")
  }else{
    main = paste0("generation ", gen)
  }
  mtext(main, outer=TRUE)
  
  points(df_sub$inv_mid, df_sub$inv_FST, type="h", lwd=2, col="cornflowerblue")
  
  for (i in 1:nrow(df_sub)){
    #points(x = c(df_sub$inv_pos[i], df_sub$inv_end[i]),
     #     y = c(-0.005, -0.005), 
    #      type = "l", lwd=5, 
    #      col=rgb(cr(df_sub$freq[i]), max=255)
    arrows(x0=df_sub$inv_pos[i], x1=df_sub$inv_end[i],
           y0=-0.005, y1=-0.005, angle=90, length=0.05,
           code=3, 
           col=rgb(cr(df_sub$freq[i]), max=255)
           )
  }
  
  leg_args <- seq(0,1,0.1)
  legend(2e06, 0.25, legend = leg_args, adj=0,
         fill=rgb(cr(leg_args),max=255), bty="n", title = "Freq.")

 
  ### Plot 2 ####
 plot(y=c(-20,400),x=c(0, 2.3e+06) , col=rgb(0,0,0,0),
       ylab="Number QTNs within inversion", bty="l", xlab="Inversion Position")
  abline(h=0)
  
  df_sub$age <- (gen - df_sub$inv_originGen)/50000
   
  points(df_sub$inv_mid, df_sub$num_qtns, type="h", 
         lwd=2, col=rgb(cr_age(df_sub$age), max=255))
   
  for (i in 1:nrow(df_sub)){
    #points(x = c(df_sub$inv_pos[i], df_sub$inv_end[i]),
     #     y = c(-0.005, -0.005), 
    #      type = "l", lwd=5, 
    #      col=rgb(cr(df_sub$freq[i]), max=255)
    arrows(x0=df_sub$inv_pos[i], x1=df_sub$inv_end[i],
           y0=-10, y1=-10, angle=90, length=0.05,
           code=3, 
           col=rgb(cr(df_sub$freq[i]), max=255)
           )
  }
  
  leg_args <- seq(0,50000, 10000)
  legend(2e06, 400, legend = leg_args, adj=0,
         fill=rgb(cr_age(leg_args/50000),max=255), bty="n", title = "Age")
 
   mtext("Inversion Position", side=1, outer=TRUE
   )
    dev.off()
}

df_sub[df_sub$age>0.5,]
```

## Movie 2
In this movie, each chrom is plotted separately and the inversions are colored by FST, so you can see the size of them relative to the chromosome.
When I get the QTNs I'll add them to the plot.
```{r}
head(df)

df$inv_FST[df$inv_FST<0] <- 0

#k=0
length(all_gen)
for (gen in sort(all_gen)){
  #k=k+1
  df_sub <- df %>% filter(sim_gen==gen) %>% arrange(inv_pos)
  df_sub
  
   png(paste0("movie2/invmovie2_", sprintf("%05.f", gen), ".png"), width=8, height = 8,units = "in", res = 600)
    
  par(mfrow=c(1,1), mar=c(4,4,0,0), oma=c(0,0,3,0))
   ### Plot 1 ####
  plot(y=c(-0.01,22),x=c(-10, 120000) , col=rgb(0,0,0,0),
        bty="n", yaxt="n", xlab="Position", ylab=""
         )
  
  if (gen<10000){
    main = paste0("generation ", gen, " (burnin)")
  }else{
    main = paste0("generation ", gen)
  }
  mtext(main, outer=TRUE)
  
  for (k in seq(0,20,by=2)){
    polygon(c(0,100000,100000,0), c(k+0.01,k+0.01,k+0.99,k+0.99), col=adjustcolor("green", 0.2), border = NA)
  }
  

  for (i in 1:nrow(df_sub)){
    #points(x = c(df_sub$inv_pos[i], df_sub$inv_end[i]),
     #     y = c(-0.005, -0.005), 
    #      type = "l", lwd=5, 
    #      col=rgb(cr(df_sub$freq[i]), max=255)
    
    if (df_sub$chrom[i]==0){
      inv_start <- df_sub$inv_pos[i]
      inv_end <- df_sub$inv_end[i]
    }else{
      inv_start <- df_sub$inv_pos[i]%%((df_sub$chrom[i])*100000)
      inv_end <- df_sub$inv_end[i]%%((df_sub$chrom[i])*100000)
    }
    
    #r <- rnorm(nrow(df_sub), 0, sd=0.05)
    arrows(x0=inv_start, x1=inv_end,
           y0=df_sub$chrom_dec[i], y1=df_sub$chrom_dec[i], angle=70, length=0.05,
           code=3, lwd=4,
           col=rgb(cr_age(df_sub$inv_FST[i]/0.25), max=255)
           )
  }
  
  text(rep(-10,20),0:20+0.5,labels = 1:21)
  
  leg_args <- seq(0,0.25, 0.05)
  legend(100000, 20, legend = leg_args, adj=0,
         fill=rgb(cr_age(leg_args/0.25),max=255), bty="n", title = "Fst")
 
  dev.off()
  }
```


## Movie 3
In this movie, each chrom AND population is plotted separately and the inversions are colored by their frequency in that population, so you can see the size of them relative to the chromosome.
When I get the QTNs I'll add them to the plot.
```{r}
head(df)

df$inv_FST[df$inv_FST<0] <- 0

#k=0
length(all_gen)
for (gen in sort(all_gen)){
  #k=k+1
  df_sub <- df %>% filter(sim_gen==gen) %>% arrange(inv_pos)
  df_sub
  
   png(paste0("movie3/invmovie3_", sprintf("%05.f", gen), ".png"), width=8, height = 8,units = "in", res = 600)
    
  par(mfrow=c(1,1), mar=c(4,4,0,0), oma=c(0,0,3,0))
   ### Plot 1 ####
  plot(y=c(-0.01,22),x=c(-10, 220000) , col=rgb(0,0,0,0),
        bty="n", yaxt="n", xlab="Position", ylab=""
         )
  
  if (gen<10000){
    main = paste0("generation ", gen, " (burnin)")
  }else{
    main = paste0("generation ", gen)
  }
  mtext(main, outer=TRUE)
  
  for (k in seq(0,20,by=2)){
    polygon(c(0,100000,100000,0), c(k+0.01,k+0.01,k+0.99,k+0.99), col=adjustcolor("green", 0.2), border = NA)
    
    polygon(c(100000,200000,200000,100000), c(k+0.01,k+0.01,k+0.99,k+0.99), col=adjustcolor("blue", 0.2), border = NA)
  }
  

  for (i in 1:nrow(df_sub)){
    #points(x = c(df_sub$inv_pos[i], df_sub$inv_end[i]),
     #     y = c(-0.005, -0.005), 
    #      type = "l", lwd=5, 
    #      col=rgb(cr(df_sub$freq[i]), max=255)
    
    if (df_sub$chrom[i]==0){
      inv_start <- df_sub$inv_pos[i]
      inv_end <- df_sub$inv_end[i]
    }else{
      inv_start <- df_sub$inv_pos[i]%%((df_sub$chrom[i])*100000)
      inv_end <- df_sub$inv_end[i]%%((df_sub$chrom[i])*100000)
    }
    
    
    #r <- rnorm(nrow(df_sub), 0, sd=0.05)
    arrows(x0=inv_start, x1=inv_end,
           y0=df_sub$chrom_dec[i], y1=df_sub$chrom_dec[i], angle=70, length=0.05,
           code=3, lwd=4,
           col=rgb(cr_pop1(df_sub$freq_p1[i]), max=255)
           )
    
    arrows(x0=inv_start+100000, x1=inv_end+100000,
           y0=df_sub$chrom_dec[i], y1=df_sub$chrom_dec[i], angle=70, length=0.05,
           code=3, lwd=4,
           col=rgb(cr_pop2(df_sub$freq_p2[i]), max=255)
           )
  }
  
  text(rep(-10,20),0:20+0.5,labels = 1:21)
  
  leg_args <- seq(0,1, 0.2)
  legend(200000, 20, legend = leg_args, adj=0,
         fill=rgb(cr_pop1(leg_args),max=255), bty="n", title = "pop1 freq")
  
  leg_args <- seq(0,1, 0.2)
  legend(200000, 10, legend = leg_args, adj=0,
         fill=rgb(cr_pop2(leg_args),max=255), bty="n", title = "pop2 freq")
 
  dev.off()
  }
```
